draft-ietf-quic-transport.md

```diff
@@ -2352,171 +2361,136 @@ shown in the following figure and described below.
-"half-closed (local)".  When an endpoint receives all stream data a FIN flag the
-stream state becomes "half-closed (remote)".  An endpoint MUST NOT consider the
-stream state to have changed until all data has been sent, received or
-discarded.
+A RST_STREAM frame, or a STREAM frame with the FIN flag set also causes a stream
+to become "closed" immediately afterwards.
 
-A RST_STREAM frame on an "open" stream causes the stream to become
-"half-closed".  A stream that becomes "open" as a result of sending or receiving
-RST_STREAM immediately becomes "half-closed".  Sending a RST_STREAM frame causes
-the stream to become "half-closed (local)"; receiving RST_STREAM causes the
-stream to become "half-closed (remote)".
+Note:
 
-Any frame type that mentions a stream ID can be sent in this state.
+: An endpoint should not need to send a RST_STREAM frame on an "idle" stream,
+  unless specifically mandated by the application protocol.
 
 
-### half-closed (local)
+### The "open" State
 
-A stream that is in the "half-closed (local)" state MUST NOT be used for sending
-on new STREAM frames.  Retransmission of data that has already been sent on
-STREAM frames is permitted.  An endpoint MAY also send MAX_STREAM_DATA and
-RST_STREAM in this state.
+A stream in the "open" state is used for transmission of data in STREAM frames.
 
-An endpoint that closes a stream MUST NOT send data beyond the final offset that
-it has chosen, see {{state-closed}} for details.
+The sending peer can send STREAM, RST_STREAM and STREAM_BLOCKED frames in this
+state.  A receiving peer sends MAX_STREAM_DATA frames.
 
-A stream transitions from this state to "closed" when a STREAM frame that
-contains a FIN flag is received and all prior data has arrived, or when a
-RST_STREAM frame is received.
+In this state the sending endpoint MUST observe the flow control limits
+advertised by the receiving endpoint (see {{flow-control}}).
 
-An endpoint can receive any frame that mentions a stream ID in this state.
-Providing flow-control credit using MAX_STREAM_DATA frames is necessary to
-continue receiving flow-controlled frames.  In this state, a receiver MAY ignore
-MAX_STREAM_DATA frames for this stream, which might arrive for a short period
-after a frame bearing the FIN flag is sent.
+Opening a stream causes all lower-numbered streams in the same direction to
+become open.  Endpoints open streams in increasing numeric order,
+but loss or reordering can cause packets that open streams to arrive out of
+order.
 
+From the "open" state, the sending endpoint causes the stream to become "closed"
+by sending a RST_STREAM frame, or by sending a STREAM frame with the FIN flag
+set after sending all preceding data.
 
-### half-closed (remote) {#state-hc-remote}
+A stream becomes "closed" at the receiving endpoint when all data is received
+and a STREAM frame with a FIN flag is received or when a RST_STREAM frame is
+received.
```

Comment:
It'ts not clear to me if the semantics are that the receiver can expect data to be delivered up the RST_FRAME offset, or the transmitter can just decide to cancel at that offset whether data before that was ACK'ed. The formulation of the RST_STREAM suggests the receiver should not wait around, but RST_STREAM could also be used to send half a file and use RST_STREAM instead of FIN, even if this is probably not the design intent.